gh-106290: Fix edge cases around uops

gvanrossum opened the pull request:
A bit of an omnibus PR, alas (let me know if you want it split up). This does roughly the following (though not in that order):

- Tweak uops debugging output
- Fix the bug from gh-106290
- Rename `SET_IP` to `SAVE_IP` (per https://github.com/faster-cpython/ideas/issues/558)
- Add a `SAVE_IP` uop at the start of the trace (ditto)
- Allow `unbound_local_error`; this gives us uops for `LOAD_FAST_CHECK`, `LOAD_CLOSURE`, and `DELETE_FAST`
- Longer traces
- Support `STORE_FAST_LOAD_FAST`, `STORE_FAST_STORE_FAST`
- Add deps on pycore_uops.h to Makefile(.pre.in)

gvanrossum committed:
Add commented-out prints about why an op is not viable

gvanrossum committed:
Misc uops changes

- Longer traces
- Explain reason for ending a trace
- Support STORE_FAST_LOAD_FAST, STORE_FAST_STORE_FAST
- Add deps on pycore_uops.h

gvanrossum committed:
There's no reason to forbid unbound_local_error

This gives us LOAD_FAST_CHECK, LOAD_CLOSURE, and DELETE_FAST.

gvanrossum committed:
Add a SET_IP uop at the top of a trace

This doesn't work right yet, alas.
If an error occurs in the first uop, the lineno is -1.

gvanrossum committed:
Fix gh-106290 (wrong line numbers in traceback)

The trick is that EXIT_TRACE must decrement prev_instr.

gvanrossum committed:
Rename SET_IP to SAVE_IP

To match https://github.com/faster-cpython/ideas/issues/558.

gvanrossum committed:
Also fix prev_instr when deoptimizing

gvanrossum committed:
TEMPORARY -- turn on uops always

gvanrossum committed:
Revert "TEMPORARY -- turn on uops always"

(I just did that to be able to benchmark it and run the tests.)

This reverts commit 5b85a49d643f5d9966400fdf73b8f0d8d3c301e3.

gvanrossum committed:
Streamline debug output in optimizer.c

gvanrossum committed:
Streamline debug output in ceval.c

brandtbucher commented on Python/optimizer.c:
```diff
@@ -388,49 +404,45 @@ translate_bytecode_to_trace(
+    return 0;
 
 #undef ADD_TO_TRACE
+#undef ADD_TO_TRACE_BASIC
```
I don't see where this is defined...

gvanrossum replied:
Turd from the last cleanup. I'll get rid of it, then land it.

brandtbucher reviewed (approved):
The code itself looks good, haven't tried running it myself or anything. Just one note:

gvanrossum committed:
Fix stray #undef

Use it to `#undef DPRINTF` instead (which also depends on a variable in this scope).

gvanrossum committed:
Merge branch 'python:main' into tweak-uops
